### PR TITLE
Handle exceptions thrown in child processes forked by ProcessManager

### LINKED ISSUE #30622: Process Manager doesn't handle exceptions properly in forked processes
The following bug can have fatal consequences if somebody employed any form of exception handling around indexers in their installation, as it could lead to very hard to discover bugs, that appear only on environments with several threads available, and MAGE_INDEXER_THREADS_COUNT env variable set. 
### Preconditions (*)

1. Magento 2.3.0 CE/EE or higher
2. Magento 2.4-develop
3. Several stores are active
4. (Optional) Elasticsearch is configured as search_engine - for ease of reproducing

### Steps to reproduce (*)

1. Run `MAGE_INDEXER_THREADS_COUNT=4 php bin/magento indexer:reindex catalogsearch_fulltext`
2. Kill Elasticsearch process to simulate an exception thrown

### Expected result (*)

1. Main process exits and the message is displayed only once, the exceptions from the forked processes are not handled as they were thrown in the main process.

### Actual result (*)

1. Observe multiple Indexer failure logs and messages
```
xx@xx:/var/www/magento2$ MAGE_INDEXER_THREADS_COUNT=4 php bin/magento indexer:reindex catalogsearch_fulltext
Catalog Search indexer process unknown error:
Indexer handler is not available: elasticsearch7
Catalog Search indexer process unknown error:
Indexer handler is not available: elasticsearch7
Catalog Search indexer process unknown error:
Indexer handler is not available: elasticsearch7
Catalog Search indexer process unknown error:
Indexer handler is not available: elasticsearch7
Catalog Search indexer process unknown error:
Catalog Search indexer process unknown error:
Catalog Search indexer process unknown error:
Indexer handler is not available: elasticsearch7
Indexer handler is not available: elasticsearch7
Indexer handler is not available: elasticsearch7
Catalog Search indexer process unknown error:
Fail in child process
```
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [x] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by pawel-siejba**
### Description (*)

The following bug can have fatal consequences if somebody employed any form of exception handling around indexers in their installation, as it could lead to very hard to discover bugs, that appear only on environments with several threads available, and MAGE_INDEXER_THREADS_COUNT env variable set.

As of now in Vanilla Magento 3 indexers(catalogsearch_fulltext, catalog_category_product, catalog_product_price) use the process manager for splitting index dimensions into separate processes.

If one of the subprocesses fails, it's not exited immediately, instead the exception continues through parent's execution flow

### Fixed Issues (if relevant)

1. Fixes magento/magento2#30622

### Manual testing scenarios (*)

1. Run MAGE_INDEXER_THREADS_COUNT=4 php bin/magento indexer:reindex catalogsearch_fulltext
2. Kill Elasticsearch process to simulate an exception thrown or hard code an exception in the callables passed to ProcessManager

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by pawel-siejba** [comment]
Can be easily back ported all the way to 2.3.0 Without any changes up to 2.3.2, 2.3.0 - 2.3.1 have some phpcs comments removed, but the class stays intact.

**Comment by ihor-sviziev** [comment]
@pawel-siejba 
Your changes looks good. Let's wait for test results.
Will you be able to cover your change with any type of automated tests?

**Comment by pawel-siejba** [comment]
@magento run Static Tests

**Comment by pawel-siejba** [comment]
@ihor-sviziev I'll try to write some Unit Tests but I'm not sure if PHPUnit will handle that well as we will be essentially testing process forking.

**Comment by pawel-siejba** [comment]
@magento run all tests

**Comment by pawel-siejba** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by pawel-siejba** [comment]
@ihor-sviziev It doesn't look like writing Unit tests in this case will be feasible. Additional extensions installed may be required,  on top of that it looks to be impossible to Unit test this implementation of Process manager in a way that would prove useful to this PR - without mocking/replacing the pcntl extension function calls. Moreover forking inside a Unit test doesn't seem like the best idea to me.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8375](https://jira.corp.magento.com/browse/ENGCOM-8375) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
:x: **Before:**
#30626Main

:heavy_check_mark: **After:**
#30626Pr

**Comment by engcom-Delta** [comment]
**Note:** Functional Tests B2B  and Functional Tests CE tests are failed

File: app/code/Magento/Indexer/Model/ProcessManager.php [hunk 1]
```diff
@@ -135,11 +148,20 @@ private function isSetupMode(): bool
...
+        try {
+            // phpcs:ignore Magento2.Functions.DiscouragedFunction
+            $status = call_user_func($userFunction);
+            $status = is_int($status) ? $status : 0;
+        } catch (\Throwable $e) {
+            $status = 1;
+            $this->logger->error(
+                __('Child process failed with message: %1', $e->getMessage()),
+                $e->getTrace()
+            );
```
Comment (ihor-sviziev) [inline app/code/Magento/Indexer/Model/ProcessManager.php]: Just found that 2nd argument should be array, also sending whole stack trace as an array is quite useless thing. It's better to use something like this
```suggestion
            $this->logger->error(
                __('Child process failed with message: %1', $e->getMessage()),
                ['exception' => (string)$e]
            );
```

Or 
```suggestion
            $this->logger->error(
                __('Child process failed with message: %1', $e->getMessage()),
                ['exception_trace' => $e->getTraceAsString()]
            );
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by pawel-siejba** [comment]
@magento run all tests

File: app/code/Magento/Indexer/Model/ProcessManager.php [hunk 2]
```diff
@@ -135,11 +148,20 @@ private function isSetupMode(): bool
...
     private function startChildProcess(callable $userFunction)
     {
-        // phpcs:ignore Magento2.Functions.DiscouragedFunction
-        $status = call_user_func($userFunction);
-        $status = is_int($status) ? $status : 0;
-        // phpcs:ignore Magento2.Security.LanguageConstruct.ExitUsage
-        exit($status);
+        try {
+            // phpcs:ignore Magento2.Functions.DiscouragedFunction
+            $status = call_user_func($userFunction);
```
Comment (ihor-sviziev) [inline app/code/Magento/Indexer/Model/ProcessManager.php:153]: Can we remove call_user_func here? Like this
```suggestion
            $status = $userFunction();
```

Comment (pawel-siejba) [reply app/code/Magento/Indexer/Model/ProcessManager.php:153]: Wouldn't be really backwards compatible as `call_user_func` and therefore `\Magento\Indexer\Model\ProcessManager::execute` can also have strings passed in the form of `\Foo::test`, which would throw an exception in the suggested case.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8375](https://jira.corp.magento.com/browse/ENGCOM-8375) has been created to process this Pull Request

**Review by engcom-Delta — CHANGES_REQUESTED** [review]
@pawel-siejba  with new changes we have same output as before but it is recorded into exception.log file. Could you clarify if this is expected behavior with your fix?
**Manual testing scenario:**
- Create additional website with products
image
- Run MAGE_INDEXER_THREADS_COUNT=4 php bin/magento indexer:reindex catalogsearch_fulltext
- Kill Elasticsearch process to simulate an exception thrown (e.g., `kill pid` command)

**Output on 2.4-develop:**
Screenshot from 2020-11-17 21-40-26

**Output with fix in console:**
image
**Output with fix in log file:**
[exception.log](https://github.com/magento/magento2/files/5559035/exception.log)

**Comment by pawel-siejba** [comment]
@engcom-Delta Yes, this is the expected behaviour. BTW don't know if you noticed but you unveiled another bug with the ProcessManager that I will create a separate issue for as it's not related to this change and also reproducible without this change. Basically if MAGE_INDEXER_THREADS_COUNT > Number of dimensions(in your case 3 stores) the main process always exits successfully.

**Comment by engcom-Delta** [comment]
@pawel-siejba  thanks for clarification
@ihor-sviziev Could you approve to move PR into ready for testing column?

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8375](https://jira.corp.magento.com/browse/ENGCOM-8375) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
Was checked case from Manual testing scenarios
**Before:**
:x: Errors are shown in the console
image


**After:**
:heavy_check_mark: Errors are recorded into log file
image
exception.log:
```
[2020-11-19 16:24:48] main.ERROR: Child process failed with message: <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>405 Method Not Allowed</title>
</head><body>
<h1>Method Not Allowed</h1>
<p>The requested method PUT is not allowed for this URL.</p>
<hr>
<address>Apache/2.4.41 (Ubuntu) Server at localhost Port 80</address>
</body></html>
 {"exception":"[object] (Elasticsearch\\Common\\Exceptions\\BadRequest400Exception(code: 405): <!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">
<html><head>
<title>405 Method Not Allowed</title>
</head><body>
<h1>Method Not Allowed</h1>
<p>The requested method PUT is not allowed for this URL.</p>
<hr>
<address>Apache/2.4.41 (Ubuntu) Server at localhost Port 80</address>
</body></html>
 at /var/www/html/magento2dev/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:632)"} []
[2020-11-19 16:24:48] main.ERROR: Child process failed with message: <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>405 Method Not Allowed</title>
</head><body>
<h1>Method Not Allowed</h1>
<p>The requested method PUT is not allowed for this URL.</p>
<hr>
<address>Apache/2.4.41 (Ubuntu) Server at localhost Port 80</address>
</body></html>
 {"exception":"[object] (Elasticsearch\\Common\\Exceptions\\BadRequest400Exception(code: 405): <!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">
<html><head>
<title>405 Method Not Allowed</title>
</head><body>
<h1>Method Not Allowed</h1>
<p>The requested method PUT is not allowed for this URL.</p>
<hr>
<address>Apache/2.4.41 (Ubuntu) Server at localhost Port 80</address>
</body></html>
 at /var/www/html/magento2dev/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:632)"} []
```

**Note:** Issue with the main process always exits successfully is reproducible on 2.4-develop without fix:
https://github.com/magento/magento2/pull/30626#issuecomment-729598126

**Comment by engcom-Delta** [comment]
**Note:** Automation tests are passed

**Pull request merged**